File: store-web/src/lib/api/product.ts
import type { ProductDto } from "$lib/dto/product";
import { restController } from "./controller";

export type createProductParams = {
  name: string;
  description: string;
  tags: string[];
  preview: string;
  price: number;
  characteristics: string;
}

export type createProductResponse = {
  path: string;
}

export type getProductParams = {
  path: string;
}

export type getProductFeedParams = {
  id: number;
}

export type searchProductsParams = {
  page: number | null;
  tags: string[] | string | null;
  text: string | null;
  authorID: number | null;
}

export async function createProduct(params: createProductParams)  {
  return restController.authCall<createProductResponse>({
    path: "/product/new",
    method: "POST",
    body: params
  })
}

export async function getProduct(params: getProductParams): Promise<ProductDto> {
  return restController.call<ProductDto>({
    path: `/product/${params.path}`,
    method: "GET"
  })
}

export async function getProductFeed(params: getProductFeedParams): Promise<ProductDto[]> {
  return restController.call<ProductDto[]>({
    path: `/feed_content?page=${params.id}`,
    method: "GET"
  })
}

export async function searchProduct(params: searchProductsParams): Promise<ProductDto[]> {
  return restController.call<ProductDto[]>({
    path: `/search_product?page=${params.page}&tags[]=${params.tags?.toString() ? params.tags?.toString():"" }&text=${params.text ? params.text : ""}`,
    method: "GET"
  })
}

File: store-api/dto/product.go
package dto

import (
	"blog-api/model"
)

type Product struct {
	ID              int      `json:"id"`
	Name            string   `json:"name"`
	Path            string   `json:"path"`
	Description     string   `json:"description"`
	Tags            []string `json:"tags"`
	Price           int      `json:"price"`
	Preview         string   `json:"preview"`
	Characteristics string   `json:"characteristics"`
}

func NewProduct(a model.Product) Product {
	return Product{
		ID:              a.ID,
		Name:            a.Name,
		Path:            a.Path,
		Description:     a.Description,
		Tags:            a.Tags,
		Preview:         a.Preview,
		Price:           a.Price,
		Characteristics: a.Characteristics,
	}
}

func NewProducts(a []model.Product) []Product {
	articles := []Product{}
	for i := range a {
		articles = append(articles, NewProduct(a[i]))
	}
	return articles
}
